transports/quic/src/lib.rs

```diff
@@ -0,0 +1,653 @@
+                                        SocketAddr::new(ip, self.socket_addr().port());
+                                    let ma = socketaddr_to_multiaddr(&socket_addr);
+                                    tracing::debug!("New listen address: {}", ma);
+                                    return Some(TransportEvent::NewAddress {
+                                        listener_id: self.listener_id,
+                                        listen_addr: ma,
+                                    });
+                                }
+                            }
+                            Ok(IfEvent::Down(inet)) => {
+                                let ip = inet.addr();
+                                if self.socket_addr().is_ipv4() == ip.is_ipv4() {
+                                    let socket_addr =
+                                        SocketAddr::new(ip, self.socket_addr().port());
+                                    let ma = socketaddr_to_multiaddr(&socket_addr);
+                                    tracing::debug!("Expired listen address: {}", ma);
+                                    return Some(TransportEvent::AddressExpired {
+                                        listener_id: self.listener_id,
+                                        listen_addr: ma,
+                                    });
+                                }
+                            }
+                            Err(error) => {
+                                tracing::debug! {
+                                    "Failure polling interfaces: {:?}.",
+                                    error
+                                };
+                                return Some(TransportEvent::ListenerError {
+                                    listener_id: self.listener_id,
+                                    error,
+                                });
+                            }
+                        }
+                    }
+                }
+                Poll::Pending => return None,
+            }
+        }
+    }
+}
+
+impl Stream for Listener {
+    type Item = TransportEvent<<QuicTransport as Transport>::ListenerUpgrade, io::Error>;
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+        if let Some(closed) = this.report_closed.as_mut() {
+            // Listener was closed.
+            // Report the transport event if there is one. On the next iteration, return
+            // `Poll::Ready(None)` to terminate the stream.
+            return Poll::Ready(closed.take());
+        }
+        if let Some(event) = this.poll_if_addr(cx) {
+            return Poll::Ready(Some(event));
+        }
+        let connecting = match futures::ready!(this.accept.poll_unpin(cx)) {
+            Some(c) => {
+                let endpoint = this.endpoint.clone();
+                this.accept = Box::pin(async move { endpoint.accept().await });
```

Comment:
What is the status on `quinn` exposing named futures?

Reply:
They are in the `main` branch.

Reply:
Cool! That `Future` seems cancellation safe so I think we could simplify this by just saying:

```rust
this.endpoint().accept().poll_unpin(cx)
```

Reply:
Ah damn, it looks like `Notified` is actually not cancellation-safe so we have to store this future in here which unfortunately requires us to box it then.